modules/tensor_mechanics/src/materials/RadialReturnStressUpdate.C

```diff
@@ -385,12 +451,64 @@ RadialReturnStressUpdateTempl<false>::updateStateSubstep(RankTwoTensor & strain_
 
   // update stress
   stress_new = sub_stress_new;
+
   // update effective inelastic strain
   _effective_inelastic_strain[_qp] =
       _effective_inelastic_strain_old[_qp] + sub_scalar_effective_inelastic_strain;
+}
+
+template <>
+void
+RadialReturnStressUpdateTempl<false>::updateStateSubstep(RankTwoTensor & strain_increment,
+                                                         RankTwoTensor & inelastic_strain_increment,
+                                                         const RankTwoTensor & rotation_increment,
+                                                         RankTwoTensor & stress_new,
+                                                         const RankTwoTensor & stress_old,
+                                                         const RankFourTensor & elasticity_tensor,
+                                                         const RankTwoTensor & elastic_strain_old,
+                                                         bool compute_full_tangent_operator,
+                                                         RankFourTensor & tangent_operator)
+{
+  unsigned int num_substeps = calculateNumberSubsteps(strain_increment);
+  while (true)
+  {
+    try
+    {
+      updateStateSubstepInternal(strain_increment,
+                                 inelastic_strain_increment,
+                                 rotation_increment,
+                                 stress_new,
+                                 stress_old,
+                                 elasticity_tensor,
+                                 elastic_strain_old,
+                                 num_substeps,
+                                 compute_full_tangent_operator,
+                                 tangent_operator);
+    }
+    catch (MooseException & e)
+    {
+      // if we are not using adaptive substepping we just rethrow the exception
+      if (!_adaptive_substepping)
+        throw e;
+
+      // otherwise we double the number of substeps and try again
+      num_substeps *= 2;
+      if (num_substeps <= _maximum_number_substeps)
```

Comment:
We would effectively never reach anywhere closer to the maximum number of substeps. Not a big deal, but the user may think the maximum number of substeps was, at some point, reached. I'd leave it as it is though.

Reply:
> We would effectively never reach anywhere closer to the maximum number of substeps.

Why is that?

Reply:
Maybe I am missing something according to your reply. The way I read it is: User selects maximum number of substeps = 7. Then, we have 1 substeps, failure, two substeps, failure, four substeps, failure, we try to make 8 substeps, but it's not less than or equal to 7 so we got too many substeps. The user wanted 7 and got 4.

Reply:
Oh, I see. I could try with the maximum number, but that could result in case where - if the user selects a maximum of 9 substeps, the algorithm tries 1,2,4,8,9, with 9 unlikely to be a substantial improvement. And the user really doesn't have a way to anticipate either scenario, as the suggested error based number of substeps may be 3, resulting in 3,6,7 substeps in your example.

I don't have a good answer here, other than maybe rewording the doc to state "...will double the amount of substeps as long as the number of substeps stays below the user specified maximum ..."

Reply:
Yeah, maybe just add that clarifying sentence to the doc.